inotify.go

```diff
@@ -314,20 +352,65 @@ func (e *Event) ignoreLinux(mask uint32) bool {
 	return false
 }
 
+func (w *Watcher) isRecursive(path string) *watch {
+	ww, ok := w.watches[path]
+	if !ok {
+		path = filepath.Dir(path)
+		ww, ok = w.watches[path]
+		if !ok {
+			return nil
+		}
+	}
+	if !w.paths[int(ww.wd)].recurse {
+		return nil
+	}
+	return ww
+}
+
 // newEvent returns an platform-independent Event based on an inotify mask.
-func newEvent(name string, mask uint32) Event {
+func (w *Watcher) newEvent(name string, mask uint32) Event {
 	e := Event{Name: name}
 	if mask&unix.IN_CREATE == unix.IN_CREATE || mask&unix.IN_MOVED_TO == unix.IN_MOVED_TO {
 		e.Op |= Create
+
+		// Add new directories on recursive watches.
+		if mask&unix.IN_ISDIR == unix.IN_ISDIR {
+			ww := w.isRecursive(name)
+			if ww != nil {
+				//err := w.add(name, true)
+				err := w.Add(filepath.Join(name, "..."))
+				if err != nil {
+					// TODO: not sure if this has a nice error message.
+					//       Also, this path could have been removed by now;
+					//       should probably filter ENOENT or something.
+					w.Errors <- err
```

Comment:
this could block the event loop.
I'd recommend adding a select with another case for receiving from w.done, just to make sure that we're never blocking on send to the error channel after the watcher has been closed, as in:
```
			select {
			case w.Errors <- err:
			case <-w.done:
				return
			}
```